File: internal/configmerge/configmerge_test.go
package configmerge

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	"github.com/BurntSushi/toml"

	"github.com/ricardoborges-teachable/ai-setup/internal/files"
)

func TestMergeJSONFile_NewFile(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "settings.json")
	patch := map[string]any{
		"permissions": map[string]any{"allow": []any{"Read"}},
	}
	bak, err := MergeJSONFile(path, patch)
	if err != nil {
		t.Fatalf("merge: %v", err)
	}
	if bak != "" {
		t.Errorf("want empty backupPath for new file, got %q", bak)
	}
	got := readJSON(t, path)
	if !reflect.DeepEqual(got, patch) {
		t.Errorf("content mismatch: got %v, want %v", got, patch)
	}
}

func TestMergeJSONFile_PreservesUserKeys(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "settings.json")

	// Pre-author a user-owned file with keys we should not touch.
	user := map[string]any{
		"experimental": map[string]any{"foo": true},
		"permissions":  map[string]any{"allow": []any{"UserTool"}},
	}
	writeJSON(t, path, user)

	patch := map[string]any{
		"mcpServers": map[string]any{
			"orchestrator": map[string]any{"command": "ai-setup", "args": []any{"server"}},
		},
	}
	bak, err := MergeJSONFile(path, patch)
	if err != nil {
		t.Fatalf("merge: %v", err)
	}
	if bak == "" {
		t.Fatal("expected backup path on first touch")
	}
	if !files.FileExists(bak) {
		t.Fatalf("backup not created at %s", bak)
	}

	got := readJSON(t, path)
	// User keys survive.
	if got["experimental"] == nil {
		t.Errorf("user experimental key lost")
	}
	if !reflect.DeepEqual(got["permissions"], user["permissions"]) {
		t.Errorf("user permissions mutated: got %v, want %v", got["permissions"], user["permissions"])
	}
	// Patch key added.
	if got["mcpServers"] == nil {
		t.Errorf("patch mcpServers not added")
	}

	// Backup holds original, not merged result.
	bakContent := readJSON(t, bak)
	if !reflect.DeepEqual(bakContent, user) {
		t.Errorf("backup mismatch: got %v, want %v", bakContent, user)
	}
}

func TestMergeJSONFile_Idempotent(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "settings.json")
	user := map[string]any{"existing": "yes"}
	writeJSON(t, path, user)

	patch := map[string]any{"mcpServers": map[string]any{"x": map[string]any{"command": "c"}}}
	bak1, err := MergeJSONFile(path, patch)
	if err != nil {
		t.Fatalf("first merge: %v", err)
	}
	firstContent, _ := os.ReadFile(path)
	firstBak, _ := os.ReadFile(bak1)

	// Second identical run.
	bak2, err := MergeJSONFile(path, patch)
	if err != nil {
		t.Fatalf("second merge: %v", err)
	}
	secondContent, _ := os.ReadFile(path)
	secondBak, _ := os.ReadFile(bak2)

	if bak1 != bak2 {
		t.Errorf("backup path changed between runs: %q vs %q", bak1, bak2)
	}
	if string(firstContent) != string(secondContent) {
		t.Errorf("content changed on idempotent re-run")
	}
	if string(firstBak) != string(secondBak) {
		t.Errorf(".bak changed on re-run — must remain the original")
	}

	// Third run with a *different* patch: file updates, .bak unchanged.
	patch2 := map[string]any{"mcpServers": map[string]any{"y": map[string]any{"command": "d"}}}
	_, err = MergeJSONFile(path, patch2)
	if err != nil {
		t.Fatalf("third merge: %v", err)
	}
	thirdContent, _ := os.ReadFile(path)
	thirdBak, _ := os.ReadFile(bak1)
	if string(thirdContent) == string(secondContent) {
		t.Errorf("file did not update on patch change")
	}
	if string(thirdBak) != string(firstBak) {
		t.Errorf(".bak was overwritten on third run")
	}
}

func TestMergeJSONFile_SliceReplacedWholesale(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "settings.json")
	writeJSON(t, path, map[string]any{
		"permissions": map[string]any{"allow": []any{"A", "B"}},
	})
	patch := map[string]any{
		"permissions": map[string]any{"allow": []any{"C"}},
	}
	if _, err := MergeJSONFile(path, patch); err != nil {
		t.Fatalf("merge: %v", err)
	}
	got := readJSON(t, path)
	perms := got["permissions"].(map[string]any)
	allow := perms["allow"].([]any)
	if len(allow) != 1 || allow[0] != "C" {
		t.Errorf("slice not replaced wholesale: got %v, want [C]", allow)
	}
}

func TestMergeTOMLFile_NewFile(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "config.toml")
	patch := map[string]any{
		"mcp_servers": map[string]any{
			"context7": map[string]any{"command": "npx", "args": []any{"-y", "@context7/server"}},
		},
	}
	bak, err := MergeTOMLFile(path, patch)
	if err != nil {
		t.Fatalf("merge: %v", err)
	}
	if bak != "" {
		t.Errorf("want empty backupPath for new file, got %q", bak)
	}
	got := readTOML(t, path)
	if got["mcp_servers"] == nil {
		t.Fatalf("mcp_servers missing after merge: %v", got)
	}
}

func TestMergeTOMLFile_PreservesUserTable(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "config.toml")
	original := "[profile]\nname = \"me\"\n\n[profile.options]\nverbose = true\n"
	if err := os.WriteFile(path, []byte(original), 0o644); err != nil {
		t.Fatal(err)
	}
	patch := map[string]any{
		"mcp_servers": map[string]any{
			"context7": map[string]any{"command": "npx"},
		},
	}
	bak, err := MergeTOMLFile(path, patch)
	if err != nil {
		t.Fatalf("merge: %v", err)
	}
	if bak == "" {
		t.Fatal("expected backup on first touch")
	}

	got := readTOML(t, path)
	if got["profile"] == nil {
		t.Errorf("user [profile] lost after merge")
	}
	if got["mcp_servers"] == nil {
		t.Errorf("patch mcp_servers missing after merge")
	}

	// Second run with same patch: idempotent.
	firstBytes, _ := os.ReadFile(path)
	if _, err := MergeTOMLFile(path, patch); err != nil {
		t.Fatalf("second merge: %v", err)
	}
	secondBytes, _ := os.ReadFile(path)
	if string(firstBytes) != string(secondBytes) {
		t.Errorf("TOML content changed on idempotent re-run:\nfirst:\n%s\nsecond:\n%s", firstBytes, secondBytes)
	}
}

// ---------- helpers ----------

func writeJSON(t *testing.T, path string, v any) {
	t.Helper()
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, data, 0o644); err != nil {
		t.Fatal(err)
	}
}

func readJSON(t *testing.T, path string) map[string]any {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var out map[string]any
	if err := json.Unmarshal(data, &out); err != nil {
		t.Fatalf("parse %s: %v\n%s", path, err, data)
	}
	return out
}

func readTOML(t *testing.T, path string) map[string]any {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	out := map[string]any{}
	if _, err := toml.Decode(string(data), &out); err != nil {
		t.Fatalf("parse %s: %v\n%s", path, err, data)
	}
	return out
}

File: packages/ai-setup-ts/src/__tests__/configmerge.test.ts
import { mkdtempSync, readFileSync, rmSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'node:path'
import { afterEach, describe, expect, it } from 'vitest'
import { deepMerge, marshalSortedJson, mergeJsonFile } from '../utils/configmerge.js'
import { fileExists } from '../utils/files.js'

describe('configmerge', () => {
  const tempDirs: string[] = []

  const makeTempDir = (): string => {
    const dir = mkdtempSync(path.join(tmpdir(), 'ai-setup-configmerge-'))
    tempDirs.push(dir)
    return dir
  }

  const readJson = (p: string): Record<string, unknown> =>
    JSON.parse(readFileSync(p, 'utf-8')) as Record<string, unknown>

  afterEach(() => {
    for (const dir of tempDirs) {
      rmSync(dir, { recursive: true, force: true })
    }
    tempDirs.length = 0
  })

  it('writes a new file with no backup', () => {
    const dir = makeTempDir()
    const p = path.join(dir, 'settings.json')
    const patch = { permissions: { allow: ['Read'] } }

    const { backupPath } = mergeJsonFile(p, patch)

    expect(backupPath).toBeNull()
    expect(readJson(p)).toEqual(patch)
  })

  it('preserves user keys and creates .bak on first touch', () => {
    const dir = makeTempDir()
    const p = path.join(dir, 'settings.json')
    const user = {
      experimental: { foo: true },
      permissions: { allow: ['UserTool'] },
    }
    writeFileSync(p, JSON.stringify(user, null, 2))

    const patch = {
      mcpServers: { orchestrator: { command: 'ai-setup', args: ['server'] } },
    }
    const { backupPath } = mergeJsonFile(p, patch)

    expect(backupPath).not.toBeNull()
    expect(fileExists(backupPath!)).toBe(true)

    const got = readJson(p)
    expect(got.experimental).toEqual(user.experimental)
    expect(got.permissions).toEqual(user.permissions)
    expect(got.mcpServers).toEqual(patch.mcpServers)

    // .bak holds original, not merged
    expect(readJson(backupPath!)).toEqual(user)
  })

  it('is idempotent — re-run with same patch produces identical bytes; .bak never overwritten', () => {
    const dir = makeTempDir()
    const p = path.join(dir, 'settings.json')
    const user = { existing: 'yes' }
    writeFileSync(p, JSON.stringify(user, null, 2))

    const patch = { mcpServers: { x: { command: 'c' } } }
    const { backupPath: bak1 } = mergeJsonFile(p, patch)
    const firstContent = readFileSync(p, 'utf-8')
    const firstBak = readFileSync(bak1!, 'utf-8')

    const { backupPath: bak2 } = mergeJsonFile(p, patch)
    const secondContent = readFileSync(p, 'utf-8')
    const secondBak = readFileSync(bak2!, 'utf-8')

    expect(bak1).toBe(bak2)
    expect(firstContent).toBe(secondContent)
    expect(firstBak).toBe(secondBak)

    // Third run with different patch: file updates, .bak unchanged
    const patch2 = { mcpServers: { y: { command: 'd' } } }
    mergeJsonFile(p, patch2)
    const thirdContent = readFileSync(p, 'utf-8')
    const thirdBak = readFileSync(bak1!, 'utf-8')

    expect(thirdContent).not.toBe(secondContent)
    expect(thirdBak).toBe(firstBak)
  })

  it('replaces arrays wholesale (no concatenation)', () => {
    const dir = makeTempDir()
    const p = path.join(dir, 'settings.json')
    writeFileSync(p, JSON.stringify({ permissions: { allow: ['A', 'B'] } }, null, 2))

    const patch = { permissions: { allow: ['C'] } }
    mergeJsonFile(p, patch)

    const got = readJson(p)
    expect((got.permissions as { allow: string[] }).allow).toEqual(['C'])
  })

  it('recurses into nested objects; patch wins on leaf collisions', () => {
    const dir = makeTempDir()
    const p = path.join(dir, 'settings.json')
    const user = { permission: { edit: 'allow', bash: 'ask' } }
    writeFileSync(p, JSON.stringify(user, null, 2))

    const patch = { permission: { bash: 'deny' } }
    mergeJsonFile(p, patch)

    const got = readJson(p)
    expect(got.permission).toEqual({ edit: 'allow', bash: 'deny' })
  })

  it('strips JSONC comments when reading existing file', () => {
    const dir = makeTempDir()
    const p = path.join(dir, 'settings.jsonc')
    writeFileSync(p, '// user comment\n{\n  "foo": "bar"\n}\n')

    mergeJsonFile(p, { baz: 'qux' })

    const got = readJson(p)
    expect(got).toEqual({ foo: 'bar', baz: 'qux' })
  })
})

describe('deepMerge', () => {
  it('returns a new object (does not mutate inputs)', () => {
    const base = { a: 1, nested: { x: 1 } }
    const patch = { nested: { y: 2 } }

    const result = deepMerge(base, patch)

    expect(result).toEqual({ a: 1, nested: { x: 1, y: 2 } })
    expect(base).toEqual({ a: 1, nested: { x: 1 } })
    expect(patch).toEqual({ nested: { y: 2 } })
  })

  it('treats arrays as leaf values', () => {
    const result = deepMerge({ list: [1, 2] }, { list: [3] })
    expect(result.list).toEqual([3])
  })

  it('handles empty base', () => {
    const result = deepMerge({}, { a: 1 })
    expect(result).toEqual({ a: 1 })
  })
})

describe('marshalSortedJson', () => {
  it('sorts keys alphabetically at every depth', () => {
    const input = { zebra: 1, alpha: { z: 1, a: 2 }, beta: [] }
    const out = marshalSortedJson(input)

    expect(out).toBe('{\n  "alpha": {\n    "a": 2,\n    "z": 1\n  },\n  "beta": [],\n  "zebra": 1\n}\n')
  })

  it('ends with a trailing newline', () => {
    const out = marshalSortedJson({ a: 1 })
    expect(out.endsWith('\n')).toBe(true)
  })
})
